mdal/frmts/mdal_ugrid.cpp

```diff
@@ -401,3 +403,133 @@ void MDAL::DriverUgrid::parse2VariablesFromAttribute( const std::string &name, c
+  mNcFile.putAttrStr( mesh2dId, "cf_role", "mesh_topology" );
+  mNcFile.putAttrStr( mesh2dId, "long_name", "Topology data of 2D network" );
+  mNcFile.putAttrInt( mesh2dId, "topology_dimension", 2 );
+  mNcFile.putAttrStr( mesh2dId, "node_coordinates", "mesh2d_node_x mesh2d_node_y" );
+  mNcFile.putAttrStr( mesh2dId, "node_dimension", "nmesh2d_node" );
+  mNcFile.putAttrStr( mesh2dId, "max_face_nodes_dimension", "max_nmesh2d_face_nodes" );
+  mNcFile.putAttrStr( mesh2dId, "edge_node_connectivity", "mesh2d_edge_nodes" );
+  mNcFile.putAttrStr( mesh2dId, "edge_dimension", "nmesh2d_edge" );
+  mNcFile.putAttrStr( mesh2dId, "face_node_connectivity", "mesh2d_face_nodes" );
+  mNcFile.putAttrStr( mesh2dId, "face_dimension", "nmesh2d_face" );
+
+  // Nodes X coordinate
+  int mesh2dNodeXId = mNcFile.defineVar( "mesh2d_node_x", NC_DOUBLE, 1, &dimNodeCountId );
+  mNcFile.putAttrStr( mesh2dNodeXId, "units", "m" );
+  mNcFile.putAttrStr( mesh2dNodeXId, "standard_name", "projection_x_coordinate" );
+  mNcFile.putAttrStr( mesh2dNodeXId, "long_name", "x-coordinate of mesh nodes" );
+  mNcFile.putAttrStr( mesh2dNodeXId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2dNodeXId, "location", "node" );
+
+  // Nodes Y coordinate
+  int mesh2dNodeYId = mNcFile.defineVar( "mesh2d_node_y", NC_DOUBLE, 1, &dimNodeCountId );
+  mNcFile.putAttrStr( mesh2dNodeYId, "units", "m" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "standard_name", "projection_y_coordinate" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "long_name", "y-coordinate of mesh nodes" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2dNodeYId, "location", "node" );
+
+  // Nodes Z coordinate
+  int mesh2dNodeZId = mNcFile.defineVar( "mesh2d_node_z", NC_DOUBLE, 1, &dimNodeCountId );
+  mNcFile.putAttrStr( mesh2dNodeZId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "location", "node" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "coordinates", "mesh2d_node_x mesh2d_node_y" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "standard_name", "altitude" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "long_name", "z-coordinate of mesh nodes" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "units", "m" );
+  mNcFile.putAttrStr( mesh2dNodeZId, "grid_mapping", "projected_coordinate_system" );
+  mNcFile.putAttrDouble( mesh2dNodeZId, "_FillValue", -999.0 );
+
+  // Faces 2D Variable
+  int mesh2FaceNodesId_dimIds [] { dimFaceCountId, dimMaxNodesPerFaceId };
+  int mesh2FaceNodesId = mNcFile.defineVar( "mesh2d_face_nodes", NC_INT, 2, mesh2FaceNodesId_dimIds );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "cf_role", "face_node_connectivity" );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "mesh", "mesh2d" );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "location", "face" );
+  mNcFile.putAttrStr( mesh2FaceNodesId, "long_name", "Mapping from every face to its corner nodes (counterclockwise)" );
+  mNcFile.putAttrInt( mesh2FaceNodesId, "start_index", 1 );
```

Comment:
in MDAL internal arrays it is numbered from 0. So I think you either need to put 0 here or add "1" to every face/vertex index when you store it. (in case I undestand correctly this attribute...)